sdk/storage/Azure.Storage.Blobs/src/LeaseClient.cs

```diff
@@ -0,0 +1,1091 @@
+    /// The <see cref="LeaseClient"/> allows you to manipulate Azure
+    /// Storage leases on containers and blobs.
+    /// </summary>
+	public class LeaseClient
+    {
+        /// <summary>
+        /// The <see cref="BlobClient"/> to manage leases for.
+        /// </summary>
+        private readonly BlobClient _blob;
+
+        /// <summary>
+        /// The <see cref="BlobContainerClient"/> to manage leases for.
+        /// </summary>
+        private readonly BlobContainerClient _container;
+
+        /// <summary>
+        /// Gets the URI of the object being leased.
+        /// </summary>
+        public Uri Uri => this._blob?.Uri ?? this._container?.Uri;
+
+        /// <summary>
+        /// Gets the Lease ID for this lease.
+        /// </summary>
+        public string LeaseId { get; private set; }
+
+        /// <summary>
+        /// The <see cref="HttpPipeline"/> transport pipeline used to send
+        /// every request.
+        /// </summary>
+        private HttpPipeline Pipeline => this._blob?.Pipeline ?? this._container._pipeline;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LeaseClient"/>  class.
+        /// </summary>
+        protected LeaseClient()
+        {
+            this._blob = null;
+            this._container = null;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LeaseClient"/>  class.
+        /// </summary>
+        /// <param name="client">
+        /// A <see cref="BlobClient"/> representing the blob being leased.
+        /// </param>
+        /// <param name="leaseId">
+        /// An optional lease ID.  If no lease ID is provided, a random lease
+        /// ID will be created.
+        /// </param>
+        public LeaseClient(BlobClient client, string leaseId = null)
+        {
+            this._blob = client ?? throw new ArgumentNullException(nameof(client));
+            this._container = null;
+            this.LeaseId = leaseId ?? GetRandomLeaseId();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LeaseClient"/>  class.
+        /// </summary>
+        /// <param name="client">
+        /// A <see cref="BlobContainerClient"/> representing the container
+        /// being leased.
+        /// </param>
+        /// <param name="leaseId">
+        /// An optional lease ID.  If no lease ID is provided, a random lease
+        /// ID will be created.
+        /// </param>
+        public LeaseClient(BlobContainerClient client, string leaseId = null)
+        {
+            this._blob = null;
+            this._container = client ?? throw new ArgumentNullException(nameof(client));
+            this.LeaseId = leaseId ?? GetRandomLeaseId();
+        }
+
+        /// <summary>
+        /// Gets a random lease ID.
+        /// </summary>
+        /// <returns>A random lease ID.</returns>
+        private static string GetRandomLeaseId() => Guid.NewGuid().ToString();
```

Comment:
It's force of habit at this point.  Both adding the doc comments and not adding any value.  Unless I hear otherwise, I'll probably keep adding them because it's easier to add doc comments everywhere as I go than to make a large push before releasing to put them on all our public APIs.